tagger.go

```diff
@@ -65,85 +69,240 @@ func newLinodeClient() linodego.Client {
-					if !slices.Contains(rule.Tags.Absent, tag) {
-						// if this tag is not on the `absent` list,
-						// we can persist it through to the new tag set
-						newTags = append(newTags, tag)
-					}
+func compareTags(linodeObject interface{}, linodeTags []TagRule) (combinedNewTags []string, tags []string, linodeID int) {
+
+	var label string
+	switch linode := linodeObject.(type) {
+	case linodego.Instance:
+		tags = linode.Tags
+		label = linode.Label
+		linodeID = linode.ID
+	case linodego.Volume:
+		tags = linode.Tags
+		label = linode.Label
+		linodeID = linode.ID
+	case linodego.NodeBalancer:
+		tags = linode.Tags
+		label = *linode.Label
+		linodeID = linode.ID
+	case linodego.Domain:
+		tags = linode.Tags
+		label = linode.Domain
+		linodeID = linode.ID
+	case linodego.LKECluster:
+		tags = linode.Tags
+		label = linode.Label
+		linodeID = linode.ID
+	}
+
+	sort.Strings(tags)
+
+	for _, rule := range linodeTags {
+		validInstance := regexp.MustCompile(rule.Regex)
+
+		if validInstance.MatchString(label) {
+			var newTags []string
+
+			// check `absent` tags to remove unwanted tags
+			for _, tag := range tags {
+				if !slices.Contains(rule.Tags.Absent, tag) {
+					// if this tag is not on the `absent` list,
+					// we can persist it through to the new tag set
+					newTags = append(newTags, tag)
 				}
+			}
 
-				// check `present` tags to ensure specific tags exist
-				for _, tag := range rule.Tags.Present {
-					// compare filter against `newTags`, as
-					// newTags == (the linode's tags - absent tags)
-					if !slices.Contains(newTags, tag) {
-						// if the specified tag does not exist,
-						// add it into the new tag set
-						newTags = append(newTags, tag)
-					}
+			// check `present` tags to ensure specific tags exist
+			for _, tag := range rule.Tags.Present {
+				// compare filter against `newTags`, as
+				// newTags == (the linode's tags - absent tags)
+				if !slices.Contains(newTags, tag) {
+					// if the specified tag does not exist,
+					// add it into the new tag set
+					newTags = append(newTags, tag)
 				}
+			}
 
-				// add newTags to combined list of new tags for this instance,
-				// excluding duplicates
-				for _, tag := range newTags {
-					if !slices.Contains(combinedNewTags, tag) {
-						combinedNewTags = append(combinedNewTags, tag)
-					}
+			// add newTags to combined list of new tags for this instance,
+			// excluding duplicates
+			for _, tag := range newTags {
+				if !slices.Contains(combinedNewTags, tag) {
+					combinedNewTags = append(combinedNewTags, tag)
 				}
 			}
 		}
+	}
+
+	return combinedNewTags, tags, linodeID
+}
+
+func logTags(combinedNewTags []string, tags []string, linodeID int) {
+	if len(combinedNewTags) > 0 {
+		sort.Strings(combinedNewTags)
+		if !slices.Equal(tags, combinedNewTags) {
+			log.WithFields(log.Fields{
+				"linode_id": linodeID,
+				"old_tags":  tags,
+				"new_tags":  combinedNewTags,
+			}).Debug("Linode tag set updated")
+		}
+	}
+}
```

Comment:
similar comments here about not necessarily being a compute instance type anymore. 

maybe we update the log fields to `s/linode_id/id` and add a new field that denotes the actual object type? ie, `vm`, `nodebalancer`, `domain`, etc? thoughts?

Reply:
sgtm!